### PR TITLE
[PM-221]  Add Cipher ID column to Send table

### PR DESCRIPTION
**Pull request by ttalty**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Add the CipherId nullable column to the Send table and modify the create and update sprocs. 

## Code changes

* **src/Core/Tools/Entities/Send.cs:** Added CipherId to the model
* **src/Sql/Tools/dbo/Stored Procedures/Send_Create.sql:** Added CipherId to the create sproc
* **src/Sql/Tools/dbo/Stored Procedures/Send_Update.sql:** Added CipherId to the update sproc
* **src/Sql/Tools/dbo/Tables/Send.sql:** CipherId added to sql table
* **util/Migrator/DbScripts/2023-11-28_00_Send_Add_CipherId_Column.sql:** Migration script for sql table and sprocs

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Sql/Tools/dbo/Tables/Send.sql [hunk 1]
```diff
@@ -14,12 +14,13 @@
...
     [Disabled]       BIT              NOT NULL,
     [HideEmail]      BIT              NULL,
+    [CipherId]       UNIQUEIDENTIFIER NULL
     CONSTRAINT [PK_Send] PRIMARY KEY CLUSTERED ([Id] ASC),
     CONSTRAINT [FK_Send_Organization] FOREIGN KEY ([OrganizationId]) REFERENCES [dbo].[Organization] ([Id]),
     CONSTRAINT [FK_Send_User] FOREIGN KEY ([UserId]) REFERENCES [dbo].[User] ([Id])
 );
 
-
+
```
Comment (djsmith85) [inline src/Sql/Tools/dbo/Tables/Send.sql:23]: :pick: Remove this unneeded space.

File: src/Sql/Tools/dbo/Tables/Send.sql [hunk 2]
```diff
@@ -14,12 +14,13 @@
     [DeletionDate]   DATETIME2 (7)    NOT NULL,
     [Disabled]       BIT              NOT NULL,
     [HideEmail]      BIT              NULL,
+    [CipherId]       UNIQUEIDENTIFIER NULL
     CONSTRAINT [PK_Send] PRIMARY KEY CLUSTERED ([Id] ASC),
     CONSTRAINT [FK_Send_Organization] FOREIGN KEY ([OrganizationId]) REFERENCES [dbo].[Organization] ([Id]),
     CONSTRAINT [FK_Send_User] FOREIGN KEY ([UserId]) REFERENCES [dbo].[User] ([Id])
 );
```
Comment (djsmith85) [inline src/Sql/Tools/dbo/Tables/Send.sql:21]: :x: This is missing a foreign key relationship to the Cipher-Table

**Review by djsmith85 — CHANGES_REQUESTED** [review]

File: src/Sql/Tools/dbo/Tables/Send.sql (hunk 1 above)
Comment (ttalty) [reply src/Sql/Tools/dbo/Tables/Send.sql]: Whitespace removed

File: src/Sql/Tools/dbo/Tables/Send.sql (hunk 2 above)
Comment (ttalty) [reply src/Sql/Tools/dbo/Tables/Send.sql]: Foreign Key added

**Review by djsmith85 — COMMENTED** [review]
My feedback has been applied but a few workflows are failing. Will continue once these are addressed

File: src/Sql/Tools/dbo/Stored Procedures/Send_Create.sql [hunk 3]
```diff
@@ -13,7 +13,8 @@
     @ExpirationDate DATETIME2(7),
     @DeletionDate DATETIME2(7),
     @Disabled BIT,
-    @HideEmail BIT
+    @HideEmail BIT,
+    @CipherId UNIQUEIDENTIFIER
```
Comment (withinfocus) [inline src/Sql/Tools/dbo/Stored Procedures/Send_Create.sql:17]: ❌ Need a default here (`= NULL`).

File: src/Sql/Tools/dbo/Stored Procedures/Send_Create.sql [hunk 4]
```diff
@@ -34,7 +35,8 @@ BEGIN
         [ExpirationDate],
         [DeletionDate],
         [Disabled],
-        [HideEmail]
+        [HideEmail],
+        [CiperId]
```
Comment (withinfocus) [inline src/Sql/Tools/dbo/Stored Procedures/Send_Create.sql:39]: ❌ Typo, thus the validation issue.

File: src/Sql/Tools/dbo/Stored Procedures/Send_Update.sql [hunk 5]
```diff
@@ -13,7 +13,8 @@
     @ExpirationDate DATETIME2(7),
     @DeletionDate DATETIME2(7),
     @Disabled BIT,
-    @HideEmail BIT
+    @HideEmail BIT,
+    @CiperId UNIQUEIDENTIFIER
```
Comment (withinfocus) [inline src/Sql/Tools/dbo/Stored Procedures/Send_Update.sql:17]: ❌ Default and typo again.

File: src/Sql/Tools/dbo/Tables/Send.sql [hunk 6]
```diff
@@ -14,17 +14,17 @@
...
-
 GO
 CREATE NONCLUSTERED INDEX [IX_Send_UserId_OrganizationId]
     ON [dbo].[Send]([UserId] ASC, [OrganizationId] ASC);
 
 GO
 CREATE NONCLUSTERED INDEX [IX_Send_DeletionDate]
-    ON [dbo].[Send]([DeletionDate] ASC);
-
+    ON [dbo].[Send]([DeletionDate] ASC);
```
Comment (withinfocus) [inline src/Sql/Tools/dbo/Tables/Send.sql:30]: ⛏️ Newline.

File: util/Migrator/DbScripts/2023-11-28_00_Send_Add_CipherId_Column.sql [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    WHERE
+        [Id] = @Id
+
+    IF @UserId IS NOT NULL
+    BEGIN
+        EXEC [dbo].[User_BumpAccountRevisionDate] @UserId
+    END
+    -- TODO: OrganizationId bump?
+END
+GO
```
Comment (withinfocus) [inline util/Migrator/DbScripts/2023-11-28_00_Send_Add_CipherId_Column.sql:145]: ⛏️ Newline.

File: util/MySqlMigrations/Migrations/20231207194013_Send_Add_CipherId_And_ForeignKey.cs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+namespace Bit.MySqlMigrations.Migrations;
+
+/// <inheritdoc />
+public partial class Send_Add_CipherId_And_ForeignKey : Migration
+{
+    /// <inheritdoc />
+    protected override void Up(MigrationBuilder migrationBuilder)
+    {
+        migrationBuilder.AlterColumn<string>(
```
Comment (withinfocus) [inline util/MySqlMigrations/Migrations/20231207194013_Send_Add_CipherId_And_ForeignKey.cs:13]: ❌ All this is coming from a drift I think and you need to run all the latest migrations first.

**Review by withinfocus — CHANGES_REQUESTED** [review]

File: src/Sql/Tools/dbo/Tables/Send.sql (hunk 6 above)
Comment (withinfocus) [reply src/Sql/Tools/dbo/Tables/Send.sql:30]: Still a thing.

File: util/Migrator/DbScripts/2023-11-28_00_Send_Add_CipherId_Column.sql [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    @Password NVARCHAR(300),
+    @MaxAccessCount INT,
+    @AccessCount INT,
+    @CreationDate DATETIME2(7),
+    @RevisionDate DATETIME2(7),
+    @ExpirationDate DATETIME2(7),
+    @DeletionDate DATETIME2(7),
+    @Disabled BIT,
+    @HideEmail BIT,
+    @CipherId UNIQUEIDENTIFIER
```
Comment (withinfocus) [inline util/Migrator/DbScripts/2023-11-28_00_Send_Add_CipherId_Column.sql:41]: ❌ The null setting needs to be here in the script too, as this is what applies the migration.

File: util/Migrator/DbScripts/2023-11-28_00_Send_Add_CipherId_Column.sql [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
```
Comment (withinfocus) [inline util/Migrator/DbScripts/2023-11-28_00_Send_Add_CipherId_Column.sql:113]: ❌ Same thing.

File: util/Migrator/DbScripts/2023-11-28_00_Send_Add_CipherId_Column.sql (hunk 7 above)
Comment (withinfocus) [reply util/Migrator/DbScripts/2023-11-28_00_Send_Add_CipherId_Column.sql:145]: Still a thing.

File: util/MySqlMigrations/Migrations/20240102200509_Send_Add_CipherId.cs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    protected override void Up(MigrationBuilder migrationBuilder)
+    {
+        migrationBuilder.AddColumn<Guid>(
+            name: "CipherId",
+            table: "Send",
+            type: "char(36)",
+            nullable: true,
+            collation: "ascii_general_ci");
+
+        migrationBuilder.CreateTable(
```
Comment (withinfocus) [inline util/MySqlMigrations/Migrations/20240102200509_Send_Add_CipherId.cs:20]: ❌ Still out of sync. This was done in https://github.com/bitwarden/server/pull/3555.

**Review by withinfocus — CHANGES_REQUESTED** [review]
Getting close.

**Pull request closed without merging**